Capture original regex in non-matching group when an exact match is r…

libklein opened the pull request:
Fixes #7776.

Tools::convetToRegex pre-/appends "^"/"<span>$</span>" to the built regex if an exact match is requested. This breaks expressions such as "v|", which, instead of matching v or the empty string, matches v or the end of line after substitution: "^v|<span>$</span>" is parsed as "^(v|<span>$</span>)".

Surrounding the original regex with an non-capturing group fixes this error.

See https://pcre.org/pcre.txt Sections "Vertical Bar" and "Subpatterns".

## Testing strategy
Added test to TestFdoSecrets and TestTools.


## Type of change
- ✅ Bug fix (non-breaking change that fixes an issue)

michaelk83 commented on tests/TestFdoSecrets.cpp:
```diff
@@ -82,6 +82,11 @@ void TestFdoSecrets::testSpecialCharsInAttributeValue()
         QCOMPARE(res[0]->title(), QStringLiteral("titleB"));
     }
+    {
+        const auto term = Collection::attributeToTerm("testAttribute", "v|");
```
May want to test the `|v` case as well.

libklein replied:
I agree that more tests are need, but i think we should implement these in TestTools. The regex builder tests need some refactoring as well. Right now they just check if the regex built matches some expected regex, but not if they actually behave correctly.

droidmonkey commented:
Was thinking this was where the error was, great change!

Aetf commented:
Why not use [`QRegularExpression::escape`](https://doc.qt.io/qt-5/qregularexpression.html#escape-1) to escape the string
if exact match is requested? The current fix breaks on inputs like `)v|(?:`

In some sense this is pretty similar to protecting against SQL injection...

Edit: there's also [`QRegularExpression::wildcardToRegularExpression`](https://doc.qt.io/qt-5/qregularexpression.html#wildcardToRegularExpression).
But then I noticed both functions are only available in Qt 5.15...

droidmonkey commented:
That is why lol, not available yet for our minimum distros. Does the Qt function choke on these "malicious" type strings as well?

Aetf commented:
I haven't tried but I don't think so.

Here's the implementation, we should do something similar

https://code.woboq.org/qt5/qtbase/src/corelib/text/qregularexpression.cpp.html#1899

https://code.woboq.org/qt5/qtbase/src/corelib/text/qregularexpression.cpp.html#1818

droidmonkey commented:
I think I prefer just copying the 5.15 source code into our tools with attribution.

libklein commented:
I agree and will work on that and some further tests later tonight to hopefully make 2.7.1.

libklein commented:
> That is why lol, not available yet for our minimum distros. Does the Qt function choke on these "malicious" type strings as well?

Actually, the QT implementation of exact matches (https://doc.qt.io/qt-5/qregularexpression.html#anchoredPattern) does exactly the same thing were doing right now. The "malicious string" is malicious only because it is not a proper regular expression, which is something Tools::convertToRegex should require.

As we'd need the copy the existing implementation (https://github.com/qt/qtbase/blob/2eb7a92aa373f10f4e6828dad640b393279637a3/src/corelib/text/qregularexpression.cpp#L2053, available > qt 5.12) into our source anyway, i'd just leave the function as is.

I'd still implement QRegularExpression::escape (and expose it), simply because it allows us to granularize tests. See the latest commit. I'll add further tests if you agree with the approach. 

Also, does someone know if the attribution is correct?

droidmonkey commented:
Should invoke the GPL 3 statements from their license block at the top of the qregularexpression.cpp.

libklein commented:
I've fixed some more bugs with overlapping flag values, added some further tests, and the QT license header. 

After spending some time browsing through the codebase i'd suggest splitting convertToRegex into createWildcardRegex and createRegex (perhaps even get rid of the latter one) as these are orthogonal anyway.

Aetf commented on src/core/Tools.cpp:
```diff
@@ -1,7 +1,10 @@
- *  Copyright (C) 2021 KeePassXC Team <team@keepassxc.org>
+ *  Copyright (C) 2020 Giuseppe D'Angelo <dangelog@gmail.com>.
+ *  Copyright (C) 2020 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com, author Giuseppe D'Angelo
+ * <giuseppe.dangelo@kdab.com> Copyright (C) 2021 The Qt Company Ltd. Copyright (C) 2021 KeePassXC Team
```
line endings seem to be messed up

Aetf commented on src/core/Tools.cpp:
```diff
@@ -296,16 +299,80 @@ namespace Tools
+     *
+     * This function is part of the QtCore module of the Qt Toolkit. And subject to the following licenses.
+     *
+     * Commercial License Usage
+     * Licensees holding valid commercial Qt licenses may use this file in
+     * accordance with the commercial license agreement provided with the
+     * Software or, alternatively, in accordance with the terms contained in
+     * a written agreement between you and The Qt Company. For licensing terms
+     * and conditions see https://www.qt.io/terms-conditions. For further
+     * information use the contact form at https://www.qt.io/contact-us.
+     *
+     * GNU Lesser General Public License Usage
+     * Alternatively, this file may be used under the terms of the GNU Lesser
+     * General Public License version 3 as published by the Free Software
+     * Foundation and appearing in the file LICENSE.LGPL3 included in the
+     * packaging of this file. Please review the following information to
+     * ensure the GNU Lesser General Public License version 3 requirements
+     * will be met: https://www.gnu.org/licenses/lgpl-3.0.html.
```
These two shouldn't be necessary as we choose to use the GPL license?

libklein replied:
According to "https://opensource.org/faq#preserve-copyright-notices", it's best to leave the copyright as is. Im not sure if it applies to this specific situation, but i don't think it'll hurt to include everything.

droidmonkey replied:
I removed the other licenses except GPL 3.

Aetf commented on src/core/Tools.cpp:
```diff
@@ -296,16 +299,80 @@ namespace Tools
+     * will be met: https://www.gnu.org/licenses/lgpl-3.0.html.
+     *
+     * GNU General Public License Usage
+     * Alternatively, this file may be used under the terms of the GNU
```
```suggestion
     * Alternatively, this file may be used under the terms of the GNU
```
```suggestion
     * This function may be used under the terms of the GNU
```

Overall this can be confusing. What about leaving the license text, and just mentioning here copyright lines and the fact that this was adapted from Qt source.

Then in the `COPYING` file, add a new block linking to this?

libklein committed:
Correct regex escape logic

* Fixes #7776

Implement QRegularExpression::escape within Tools::convertToRegex to allow usage on older Qt versions.

Also wrap EXACT_MODIFIER patterns in a non-capture group to prevent misinterpreted regex.
